### ISSUE TITLE
`skaffold dev` stuck in infinite build loop

### ISSUE THREAD

**Issue by misterikkit** [GoogleContainerTools/skaffold#5933]
TL;DR I think `.build.artifacts.custom.dependencies.ignore` is not working as intended, causing extraneous rebuilds.

### Expected behavior
Running `skaffold dev` should only rebuild + redeploy when I make code changes.

### Actual behavior
`skaffold dev` is triggering on unexpected fsnotify events generated by the build process.

### Information

- Skaffold version: 
  v1.25.0
- Operating system: 
  Ubuntu 16.04.3 LTS
- Installed via: 
  Cloud Code
- Contents of skaffold.yaml:

```yaml
# edited for brevity. I have a dozen artifacts that are built & deployed the same way.
apiVersion: skaffold/v2beta15
kind: Config
metadata:
  name: hm
build:
  local:
    concurrency: 10
  artifacts:
    - image: protection
      context: ./protection/
      custom:
        buildCommand: ../build/go-build.sh make package PROJECT_VERSION="${IMAGE##*:}"
        dependencies:
          paths: ["."]
          ignore: ["bin/*", "deploy/chart/charts/*", "deploy/chart/crds/*", "*.pb.go", "gomock*/*", "mock/*", "pkg/agent/mock/*", "api/policy/v1/zz*"]
 
deploy:
  helm:
    releases:
      - name: protection
        skipBuildDependencies: true
        chartPath: ./protection/deploy/chart/
        artifactOverrides:
          hmlibrary.image.skaffoldOverride: protection

```

My makefile runs `mockgen`, `protoc`, and `controller-gen` as part of the build. I don't want any generated files to trigger a rebuild, since (A) that is what causes the endless rebuild loop, and (B) any inputs to the generated files _are_ part of the watched source tree. 

I see two potential problems here.
1. The dependency list is built once at the beginning, so it doesn't check the `ignore` patterns for newly created files (mockgen temp files)
2. Generated files which do exist at start time, and are matched by my `ignore` patterns (🤞) seem to make it into the file monitor event queue in an unpredictable manner.

### Steps to reproduce the behavior

1. a clonable repository with the sample skaffold project
2. `skaffold <command>`
3. ...

Skaffold logs snippet
```
Watching for changes...
time="2021-06-02T17:50:33-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/.git/index.lock\""
time="2021-06-02T17:50:33-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/.git/index.lock\""
time="2021-06-02T17:50:44-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:44-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/.git/index.lock\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/.git/index.lock\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin-go-tmp-umask\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin-go-tmp-umask\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin-go-tmp-umask\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin-go-tmp-umask\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Write: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin\""
<snip ~480 identical lines>
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.bin\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.go\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Remove: \"/home/jbasseri/dev/hm/protection/gomock_reflect_484316605/prog.go\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Write: \"/home/jbasseri/dev/hm/protection/mock/mock_policy_client.go\""
time="2021-06-02T17:50:45-06:00" level=debug msg="Change detected notify.Write: \"/home/jbasseri/dev/hm/protection/mock/mock_policy_client.go\""
time="2021-06-02T17:50:46-06:00" level=debug msg="Change detected notify.Write: \"/home/jbasseri/dev/hm/protection/mock/mock_policy_client.go\""
time="2021-06-02T17:50:46-06:00" level=debug msg="Change detected notify.Write: \"/home/jbasseri/dev/hm/protection/mock/mock_policy_client.go\""
time="2021-06-02T17:50:46-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/pkg/agent/gomock_reflect_428381803\""
time="2021-06-02T17:50:46-06:00" level=debug msg="Change detected notify.Create: \"/home/jbasseri/dev/hm/protection/pkg/agent/gomock_reflect_428381803\""
time="2021-06-02T17:50:47-06:00" level=info msg="files added: [protection/pkg/agent/gomock_reflect_428381803/prog.go]"
time="2021-06-02T17:50:47-06:00" level=info msg="files modified: [protection/mock/mock_policy_client.go]"
Generating tags...
```

The behavior seems inconsistent, as sometimes `changes.go` will log my CRD manifests as modified. e.g.
```
time="2021-06-02T18:14:03-06:00" level=info msg="files modified: [protection/deploy/chart/charts/hmlibrary-0.1.0.tgz protection/deploy/chart/crds/policy.hm.com_policies.yaml protection/deploy/chart/crds/policy.hm.com_policybindings.yaml]"
```

I believe we are hitting the default case here
https://github.com/GoogleContainerTools/skaffold/blob/fafdcd73c51cf7f95c86855c60474c50dffd7835/pkg/skaffold/sync/sync.go#L77-L78

which causes us to hit the default case here
https://github.com/GoogleContainerTools/skaffold/blob/fafdcd73c51cf7f95c86855c60474c50dffd7835/pkg/skaffold/runner/v1/dev.go#L237-L238

**Comment by misterikkit** [comment]
From my read of the skaffold code, I don't believe that leading `./` in ignore patterns should have any difference, since all those patterns go through `filepath.Join` early on. I have not confirmed this experimentally though...

**Comment by aaron-prindle** [comment]
Thanks for the detailed report here!  Do you happen to have a small app or repo setup w/ a skafflold.yaml that the skaffold team could use to reproduce this? Additionally to verify that the ignore directives are working as expected initally, can you compare the values for `a.Dependencies.Ignore`:
https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/custom/dependencies.go#L52
to see that they match what you expect?

From poking around the code I think that the `ignore:` should be re-calculated for new files as the file lister (which excludes files by the `ignore:` directives w/ glob matching) is re-run each dev-loop:
https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/list/list.go#L62-L86

To clarify, from your investigation the infinite recursion occuring because the `NewItem` event triggers a rebuild (`r.changeSet.AddRebuild(artifact)`) and this happens indefinitely for your custom build?  In that case could a potential fix be adding some case logic s.t. if a NewItem matches an `ignore:` directive a rebuild doesn't occur?

**Comment by misterikkit** [comment]
I didn't get the time to convert this into a repro app. 

> From poking around the code I think that the ignore: should be re-calculated for new files
If that's so, then the root cause is very likely user error. As mentioned above, I am running a variety of code generators across a dozen different artifacts, and it felt like I was in the infinite loop no matter what my ignore rules.

What I ended up doing was to write a script that scans my codebase for generated files and updates skaffold.yaml with patterns based on that. Once the ignore rules were systematically updated, I *think* I stopped seeing rebuild loops (currently fighting other build issues).

One thing I found particularly frustrating is that glob matching is done with `filepath.Match` which does not support shell style `**` globs. Imagine having generated protobuf code in several different packages at various depths. I ended up with things like,
```yaml
ignore:
- mock/*
- */mock/*
- */*/mock/*
```

And that's why I programatically generated my ignore rules.

I know there is also `.build.custom.dependencies.command` but I don't really want to commit a skaffold-specific script to my repo to solve this one aspect of skaffold.

**Comment by aaron-prindle** [comment]
Did the programmatic rules you generated end up working for you and stop the infinite build loop issue?  Perhaps the skaffold team should add a feature request for supporting shell style `**` globs to make this easier in the future

**Comment by misterikkit** [comment]
> Did the programmatic rules you generated end up working for you and stop the infinite build loop issue? Perhaps the skaffold team should add a feature request for supporting shell style `**` globs to make this easier in the future

Yes. I also had to add ignore rules for the temporary folders that gomock generates at various depths of my source tree (depending on where mocks get generated).

I filed the feature request in https://github.com/GoogleContainerTools/skaffold/issues/6350
